### PR TITLE
Signer heartbeat

### PR DESCRIPTION
**Pull request by amendelzon**
- Implementing heartbeat feature on the signer
- Saving the last signed BTC transaction hash upon successful authorized signing
- Implemented signer heartbeat operation in manager
- Implemented new scheme for new HSM2Dongle commands
- Implementing heartbeat feature on the TCPSigner
- Implemented attestation simulation and key/hash persistence in the TCPSigner
- Added heartbeat testing and test cases
- Added manager heartbeat unit tests
- Updated protocol specification
- Added heartbeat documentation
- Moved compile time utilities to the common directory, updated references
- Added third-party hmac-sha256 library with tests
- Added hmac-sha256 tests to CI
- Ignored additional TCPSigner-generated files

### CODE REVIEW

File: middleware/tests/ledger/hsm2dongle_cmds/test_signer_heartbeat.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    #     self.assert_exchange(exchanges)
+
+    # def test_onboard_seed_error(self):
+    #     self.dongle.exchange.side_effect = [bytes([0])]*30 + [CommException("an-error")]
+
+    #     with self.assertRaises(HSM2DongleError):
+    #         self.hsm2dongle.onboard(bytes(map(lambda i: i*2, range(32))), b"1234")
+
+    #     seed_exchanges = list(map(lambda i: [0x44, i, i*2], range(31)))
+    #     self.assert_exchange(seed_exchanges)
```
Comment (italo-sampaio) [inline middleware/tests/ledger/hsm2dongle_cmds/test_signer_heartbeat.py:126]: Are those commented-out tests supposed to be here?

**Review by italo-sampaio — CHANGES_REQUESTED** [review]
Looks great in general! A few not-mandatory style-related comments and some dead code that probably passed through by accident.

File: ledger/src/signer/src/auth.c [hunk 2]
```diff
@@ -95,6 +94,32 @@ unsigned int auth_sign(volatile unsigned int rx) {
                  APDU_DATA_PTR,
                  APDU_TOTAL_DATA_SIZE_OUT);
 
+    // Save the BTC tx hash to NVM if this signature required authorization
+    if (auth.auth_required) {
+        // Sanity check: source and destination sizes
+        if (sizeof(N_bc_state.last_auth_signed_btc_tx_hash) !=
+            sizeof(auth.tx_hash)) {
```
[the remark this answers is not in the document]
Comment (amendelzon) [reply ledger/src/signer/src/auth.c:101]: Cool, good suggestion!

File: ledger/src/signer/src/auth.h [hunk 3]
```diff
@@ -66,17 +67,14 @@
...
-    struct {
-        union {
-            uint8_t tx_hash[32];
-            uint8_t receipt_hash[32];
-        };
-        uint8_t sig_hash[32];
-    };
+    uint8_t tx_hash[32];
+    uint8_t receipt_hash[32];
+    uint8_t sig_hash[32];
```
[the remark this answers is not in the document]
Comment (amendelzon) [reply ledger/src/signer/src/auth.h:77]: Will do.

File: middleware/tests/ledger/hsm2dongle_cmds/test_signer_heartbeat.py (hunk 1 above)
Comment (amendelzon) [reply middleware/tests/ledger/hsm2dongle_cmds/test_signer_heartbeat.py:126]: Good catch, thanks!

**Review by italo-sampaio — APPROVED** [review]
Looks great! Good to go!

**Pull request merged**